docs/content/philosophy.md

```diff
@@ -78,4 +78,77 @@ dynamically into OPA remotely via APIs or through the local filesystem.
 
 OPA is a full-featured policy engine that offloads policy decisions from your
 service. You can think of it as a concierge for your service who can answer
-detailed questions on behalf of your users to meet their specific needs.
\ No newline at end of file
+detailed questions on behalf of your users to meet their specific needs.
+
+## How Does OPA Work? {#how-does-opa-work}
+
+> 📖 This section assumes you are familiar with the basics of OPA. We recommend
+> going through the [Introduction](..) before reading this section.
+
+OPA policies (written in Rego) make decisions based on hierarchical structured data.
+Sometimes we refer to this data as a document, set of attributes, piece of context,
+or even just "JSON" [1].
+
+Data can be loaded into OPA from outside world using push or pull interfaces that operate
+synchronously or asynchronously with respect to policy evaluation. We refer to all data
+loaded into OPA from the outside world as **base documents** [2]. These base documents
+almost always contribute to your policy decision-making logic. However, your policies can
+also make decisions based on each other. Policies almost always consist of multiple rules
+that refer to other rules (possibly authored by different groups). In OPA, we refer
+to the values generated by rules as **virtual documents**. The term "virtual" in this
+case just means the document is _computed_ inside the policy, i.e., it's not loaded
+into OPA from the outside world.
+
+Base and virtual documents can represent the exact same kind of information, e.g., numbers,
+strings, lists, maps, and so on. Moreover, with Rego, you can refer to both base and virtual
+documents using the exact same dot/bracket-style reference syntax. Consistency across the
+types of values that can be represented and the way those values are referenced means that
+policy authors only need to learn one way of modeling and referring to information
+that drives policy decision-making.
+
+The following table summarizes how base documents (loaded using different models)
+can be referenced by your policies and integrated into OPA.
+
+| Model | How to access in Rego | How to integrate with OPA |
+| --- | --- | --- |
+| Synchronous Push | The `input` global variable | Provide data in policy query (e.g., `POST v1/data`) |
+| Synchronous Pull | The built-in functions in OPA (e.g., `http.send`) | Provide an API that rules can query. |
+| Asynchronous Push | The `data` global variable | Invoke OPA API(s) (e.g., `PUT v1/data`)
+| Asynchronous Pull | The `data` global variable | Configure OPA's [Bundle](../management#bundles) feature. |
```

Comment:
This table feels too detailed.  I was scratching my head for a while about the Model column.  Before this table, where I  it was going was to say that all of the base and virtual documents are accessible through the API and through Rego from a single namespace: `data`.    And then continue on.  

The discussion of in-memory cache, etc. is good but feels that it should come later. once people understand how to think about /data.    Something like...
* defining "virtual docs" and "base docs" as above
* virtual/base docs unified via data.  API and Rego expressions.
* Controlling where virtual docs exist in the hierarchy via `package`
* Controlling where base docs exist in the hiearchy and the different options for loading data (e.g. this table)

Reply:
Thanks! This was super helpful. I moved the table to the end and fleshed out a few things that I thought were worthwhile:

* Mention the "arbitrary data" concept right off the bat
* Better discussion of naming/locations. This lets us answer why `input` exists for example.